<a href="https://colab.research.google.com/github/marcanoyut/Sistema_POO/blob/main/MarcanoYutzura_26110_AnalisisDatos_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#SynthData

In [1]:
from google.colab import files
uploaded = files.upload()

TypeError: 'NoneType' object is not subscriptable

##Transformación a DataFrame


In [ ]:
import pandas as pd

# leer lo que contiene ventas, marketing,cliente

df_ventas = pd.read_csv('ventas.csv')
df_clientes = pd.read_csv('clientes.csv')
df_marketing = pd.read_csv('marketing.csv')

print("¡Éxito! Los 3 sets de datos ya están en la memoria de Python.")

¡Éxito! Los 3 sets de datos ya están en la memoria de Python.


In [ ]:
# 1. Cargamos el archivo ventas (le indicamos que la columna 'fecha_venta' es una fecha)
df_ventas = pd.read_csv('ventas.csv', parse_dates=['fecha_venta'], dayfirst=True)

# 2.Limpieza del precio: quitamos el '$' y lo convertimos a número decimal (float)

df_ventas['precio'] = df_ventas['precio'].replace(r'[\$,]', '', regex=True).astype(float)
print("--- Tipos de Datos en Ventas ---")
print(df_ventas.dtypes)

--- Tipos de Datos en Ventas ---
id_venta                int64
producto               object
precio                float64
cantidad              float64
fecha_venta    datetime64[ns]
categoria              object
dtype: object


In [ ]:
df_ventas.head()

,id_venta,producto,precio,cantidad,fecha_venta,categoria
0,792,Cuadro decorativo,69.94,5.0,2024-01-02,Decoración
1,811,Lámpara de mesa,105.10,5.0,2024-01-02,Decoración
2,1156,Secadora,97.96,3.0,2024-01-02,Electrodomésticos
3,1372,Heladera,114.35,8.0,2024-01-02,Electrodomésticos
4,1546,Secadora,106.21,4.0,2024-01-02,Electrodomésticos


In [ ]:
# Corregir los encabezados

df_ventas.columns = ['ID_Venta', 'Producto', 'Precio_Unitario', 'Cantidad', 'Fecha_venta', 'Categoria']

print("Encabezados aplicados correctamente:")
df_ventas.head()

Encabezados aplicados correctamente:


,ID_Venta,Producto,Precio_Unitario,Cantidad,Fecha_venta,Categoria
0,792,Cuadro decorativo,69.94,5.0,2024-01-02,Decoración
1,811,Lámpara de mesa,105.10,5.0,2024-01-02,Decoración
2,1156,Secadora,97.96,3.0,2024-01-02,Electrodomésticos
3,1372,Heladera,114.35,8.0,2024-01-02,Electrodomésticos
4,1546,Secadora,106.21,4.0,2024-01-02,Electrodomésticos


In [ ]:
import pandas as pd


# 1. Cargamos el archivo indicando las dos columnas de fecha # Usamos dayfirst=True porque el formato es DD/MM/YYYY

df_marketing = pd.read_csv('marketing.csv', parse_dates=['fecha_inicio', 'fecha_fin'], dayfirst=True)

 # 2. Aplicamos para dar formato a encabezados

df_marketing.columns = ['ID_Campania', 'Producto', 'Canal', 'Costo', 'Fecha_Inicio', 'Fecha_Fin']

 # 3. Verificando

print("Estructura de Marketing:")
df_marketing.info()
df_marketing.head()
print("Estructura de Marketing finalizada:")









Estructura de Marketing:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   ID_Campania   90 non-null     int64         
 1   Producto      90 non-null     object        
 2   Canal         90 non-null     object        
 3   Costo         90 non-null     float64       
 4   Fecha_Inicio  90 non-null     datetime64[ns]
 5   Fecha_Fin     90 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(1), int64(1), object(2)
memory usage: 4.3+ KB
Estructura de Marketing finalizada:


In [ ]:
import pandas as pd
df_clientes = pd.read_csv('clientes.csv')

#  Aplicar encabezados
df_clientes.columns = ['ID_Cliente', 'Nombre', 'Edad', 'Ciudad', 'Ingresos']

#  Verificando
print("Base de Clientes lista:")
df_clientes.head()

# formato a la columna Ingresos
df_clientes['Ingresos'] = df_clientes['Ingresos'].astype(float)

# Verificamos el tipo de dato
print(df_clientes['Ingresos'].dtype)

df_clientes.head()




Base de Clientes lista:
float64


,ID_Cliente,Nombre,Edad,Ciudad,Ingresos
0,1,Aloysia Screase,44,Mar del Plata,42294.68
1,2,Kristina Scaplehorn,25,Posadas,24735.04
2,3,Filip Castagne,50,Resistencia,35744.85
3,4,Liuka Luard,39,Bahía Blanca,27647.96
4,5,Dore Cockshtt,28,Rosario,28245.65


##Lista de Diccionarios

In [ ]:
# 1. Convertimos los datos del DataFrame real a una Lista de Diccionarios
# Esto cumple con la consigna de usar estructuras de datos de Python puro
ventas_lista = df_ventas.to_dict('records')

# 2. Script para calcular ventas mensuales
# Usamos un diccionario para acumular los totales por mes
ventas_por_mes = {}

for venta in ventas_lista:
    # Extraemos el mes de la fecha (que ya habiamos limpiado antes)
    # .month_name() nos da el nombre del mes
    mes = venta['Fecha_venta'].month_name()

# Variables y Operadores

    cant = venta['Cantidad']
    prec = venta['Precio_Unitario']
    subtotal = cant * prec

    # Acumulación mensual
    if mes in ventas_por_mes:
        ventas_por_mes[mes] += subtotal
    else:
        ventas_por_mes[mes] = subtotal

# 4. Resultados finales
print("--- RESULTADOS MENSUALES ---")
for mes, total in ventas_por_mes.items():
    print(f"Mes: {mes:<12} | Total: ${total:,.2f}")


--- RESULTADOS MENSUALES ---
Mes: January      | Total: $nan
Mes: February     | Total: $118,672.44
Mes: March        | Total: $136,779.15
Mes: April        | Total: $144,380.10
Mes: May          | Total: $143,727.25
Mes: June         | Total: $108,480.17
Mes: July         | Total: $116,229.97
Mes: August       | Total: $119,680.15
Mes: September    | Total: $115,787.85
Mes: October      | Total: $112,117.13
Mes: November     | Total: $119,951.79
Mes: December     | Total: $117,631.94


##EDA (Exploratory Data Analysis)

In [ ]:
# Reposte de nulos y duplicados
import pandas as pd

# Creamos una lista con nuestros DataFrames y sus nombres
datasets = {
    "Ventas": df_ventas,
    "Marketing": df_marketing,
    "Clientes": df_clientes
}

print("--- REPORTE DE CALIDAD DE DATOS (ESTADO INICIAL) ---\n")

for nombre, df in datasets.items():
    print(f"FICHA TÉCNICA: {nombre}")
    print(f"Total de registros: {len(df)}")

    # Identificar Nulos
    nulos = df.isnull().sum().sum()
    print(f"Valores nulos encontrados: {nulos}")
    if nulos > 0:
        print("Detalle de nulos por columna:")
        print(df.isnull().sum()[df.isnull().sum() > 0])

    # Identificar Duplicados
    duplicados = df.duplicated().sum()
    print(f"Registros duplicados: {duplicados}")

    print("-" * 30)

--- REPORTE DE CALIDAD DE DATOS (ESTADO INICIAL) ---

FICHA TÉCNICA: Ventas
Total de registros: 3035
Valores nulos encontrados: 4
Detalle de nulos por columna:
Precio_Unitario    2
Cantidad           2
dtype: int64
Registros duplicados: 35
------------------------------
FICHA TÉCNICA: Marketing
Total de registros: 90
Valores nulos encontrados: 0
Registros duplicados: 0
------------------------------
FICHA TÉCNICA: Clientes
Total de registros: 567
Valores nulos encontrados: 0
Registros duplicados: 0
------------------------------


##  Data  Cleaning en Ventas

In [ ]:
# Eliminamos las filas duplicadas y actualizamos el DataFrame
df_ventas = df_ventas.drop_duplicates()

print(f"Duplicados eliminados. Nuevo total de registros: {len(df_ventas)}")

Duplicados eliminados. Nuevo total de registros: 3000


In [ ]:
# Eliminamos las filas que tengan nulos en estas columnas específicas
df_ventas = df_ventas.dropna(subset=['Precio_Unitario', 'Cantidad'])

print(f"Nulos eliminados. Estado final de registros: {len(df_ventas)}")

Nulos eliminados. Estado final de registros: 2998


In [ ]:
import pandas as pd

# verificamos la lista con nuestros DataFrames y sus nombres para automatizar
datasets = {
    "Ventas": df_ventas,
    "Marketing": df_marketing,
    "Clientes": df_clientes
}

print("--- REPORTE DE CALIDAD DE DATOS (ESTADO FINAL) ---\n")

for nombre, df in datasets.items():
    print(f"FICHA TÉCNICA: {nombre}")
    print(f"Total de registros: {len(df)}")

    # Identificar Nulos
    nulos = df.isnull().sum().sum()
    print(f"Valores nulos encontrados: {nulos}")
    if nulos > 0:
        print("Detalle de nulos por columna:")
        print(df.isnull().sum()[df.isnull().sum() > 0])

    # Identificar Duplicados
    duplicados = df.duplicated().sum()
    print(f"Registros duplicados: {duplicados}")

    print("-" * 30)

--- REPORTE DE CALIDAD DE DATOS (ESTADO FINAL) ---

FICHA TÉCNICA: Ventas
Total de registros: 2998
Valores nulos encontrados: 0
Registros duplicados: 0
------------------------------
FICHA TÉCNICA: Marketing
Total de registros: 90
Valores nulos encontrados: 0
Registros duplicados: 0
------------------------------
FICHA TÉCNICA: Clientes
Total de registros: 567
Valores nulos encontrados: 0
Registros duplicados: 0
------------------------------


In [ ]:
import pandas as pd

# 1. Aseguramos que los Ingresos de clientes sean float
df_clientes['Ingresos'] = df_clientes['Ingresos'].astype(float)

# 2. Pasamos Ventas a Lista de Diccionarios (cumpliendo la consigna)
ventas_lista = df_ventas.to_dict('records')

# 3. Script de Cálculo Mensual con Operadores y Variables
ventas_por_mes = {}

for venta in ventas_lista:
    # Extraemos el mes
    mes = venta['Fecha_venta'].month_name()

    # Variables y Operadores (Alineación corregida)
    # Importante: Usamos 'Cantidad' y 'Precio_Unitario' con mayúsculas como tus encabezados
    cant = venta['Cantidad']
    prec = venta['Precio_Unitario']
    subtotal = cant * prec

    # Acumulación mensual
    if mes in ventas_por_mes:
        ventas_por_mes[mes] += subtotal
    else:
        ventas_por_mes[mes] = subtotal

# 4. Resultados finales
print("--- RESULTADOS MENSUALES ---")
for mes, total in ventas_por_mes.items():
    print(f"Mes: {mes:<12} | Total: ${total:,.2f}")

--- RESULTADOS MENSUALES ---
Mes: January      | Total: $129,604.99
Mes: February     | Total: $118,672.44
Mes: March        | Total: $136,779.15
Mes: April        | Total: $128,430.69
Mes: May          | Total: $143,727.25
Mes: June         | Total: $108,480.17
Mes: July         | Total: $116,229.97
Mes: August       | Total: $119,680.15
Mes: September    | Total: $115,787.85
Mes: October      | Total: $112,117.13
Mes: November     | Total: $119,951.79
Mes: December     | Total: $117,631.94


In [ ]:
# 1. Primero, calculamos el Total de cada fila
df_ventas['Total_Venta'] = df_ventas['Precio_Unitario'] * df_ventas['Cantidad']

# 2. Definimos el umbral de "Alto Rendimiento"
# En este caso, buscaremos ventas individuales mayores a $500 (ajusta el número según tus datos)
umbral = 500

# 3. Aplicamos la Transformación (Filtro)
df_alto_rendimiento = df_ventas[df_ventas['Total_Venta'] > umbral].copy()

# 4. Ordenamos de mayor a menor para que sea más visual
df_alto_rendimiento = df_alto_rendimiento.sort_values(by='Total_Venta', ascending=False)

print(f"Se identificaron {len(df_alto_rendimiento)} registros de alto rendimiento.")
df_alto_rendimiento.head()








Se identificaron 1229 registros de alto rendimiento.


,ID_Venta,Producto,Precio_Unitario,Cantidad,Fecha_venta,Categoria,Total_Venta
550,2063,Smartphone,124.01,12.0,2024-03-05,Electrónica,1488.12
912,1109,Batidora,123.83,12.0,2024-04-11,Electrodomésticos,1485.96
866,2373,Smartphone,123.60,12.0,2024-04-07,Electrónica,1483.20
2345,2703,Smartphone,123.40,12.0,2024-10-03,Electrónica,1480.80
513,1569,Batidora,122.02,12.0,2024-02-29,Electrodomésticos,1464.24


##Agrupación por Producto

In [ ]:
# Agrupamos por producto y sumamos sus totales
ranking_productos = df_ventas.groupby('Producto')['Total_Venta'].sum().reset_index()

# Filtramos los productos que en total vendieron más de, por ejemplo, $5000
productos_top = ranking_productos[ranking_productos['Total_Venta'] > 5000]

print("Productos con alto rendimiento acumulado:")
print(productos_top)

Productos con alto rendimiento acumulado:
                  Producto  Total_Venta
0          Adorno de pared     48093.49
1                 Alfombra     44773.06
2               Aspiradora     50085.86
3              Auriculares     74175.58
4                 Batidora     50979.20
5                 Cafetera     59607.31
6               Candelabro     11128.80
7   Consola de videojuegos     46174.41
8                 Cortinas     44865.03
9        Cuadro decorativo     54297.60
10          Cámara digital     45217.96
11   Elementos de cerámica     45411.09
12       Espejo decorativo     46783.31
13      Freidora eléctrica     50155.15
14                Heladera     25736.80
15         Horno eléctrico     49913.90
16       Jarrón decorativo     51130.88
17                  Laptop     46179.48
18                Lavadora     48946.44
19         Lámpara de mesa     82276.38
20              Microondas     72562.89
21     Parlantes Bluetooth     47779.46
22        Plancha de vapor     45196.6

In [ ]:
# 1. Aseguramos que exista la columna de Total por cada fila
df_ventas['Total_Venta'] = df_ventas['Precio_Unitario'] * df_ventas['Cantidad']

# 2. Definimos un umbral: Ventas que superen el promedio
promedio_ventas = df_ventas['Total_Venta'].mean()

# 3. Creamos el nuevo DataFrame con los productos de alto rendimiento
df_alto_rendimiento = df_ventas[df_ventas['Total_Venta'] > promedio_ventas].copy()

# 4. Ordenamos para ver los más importantes arriba
df_alto_rendimiento = df_alto_rendimiento.sort_values(by='Total_Venta', ascending=False)

print(f"Umbral de rendimiento aplicado: ${promedio_ventas:.2f}")
print(f"Registros que superan el promedio: {len(df_alto_rendimiento)}")
df_alto_rendimiento.head()

Umbral de rendimiento aplicado: $489.36
Registros que superan el promedio: 1274


,ID_Venta,Producto,Precio_Unitario,Cantidad,Fecha_venta,Categoria,Total_Venta
550,2063,Smartphone,124.01,12.0,2024-03-05,Electrónica,1488.12
912,1109,Batidora,123.83,12.0,2024-04-11,Electrodomésticos,1485.96
866,2373,Smartphone,123.60,12.0,2024-04-07,Electrónica,1483.20
2345,2703,Smartphone,123.40,12.0,2024-10-03,Electrónica,1480.80
513,1569,Batidora,122.02,12.0,2024-02-29,Electrodomésticos,1464.24


## Informe de Calidad y Preparación de Datos - Etapa I y II
#Proyecto: Pre-Entrega SynthData

## Analista: Yutzura Marcano

### Hallazgos del Diagnóstico Inicial (Calidad de Datos)
Tras la auditoría realizada con **Python**, se detectaron las siguientes anomalías en el set de datos de Ventas, mientras que Marketing y Clientes se presentaron íntegros:

Registros Totales iniciales (Ventas): **3,035**

Registros Duplicados: 35 (Detectados y eliminados).

Valores Nulos: 4 registros (2 en Precio_Unitario y 2 en Cantidad). Se procedió a la eliminación por ser críticos para el cálculo.

Formato de Valores: La columna precio venía como tipo object (texto) debido al símbolo $. Los ingresos de clientes fueron normalizados a float para asegurar consistencia matemática.

###Procesamiento y Estructuras Utilizadas
Siguiendo las buenas prácticas de programación, se tomaron las siguientes decisiones técnicas:

Conversión de Tipos: Se transformaron las fechas a objetos datetime64 y los valores monetarios a float64.

Estructura de Datos: Se optó por una Lista de Diccionarios para el cálculo de ventas mensuales, facilitando la iteración y la legibilidad del código frente al uso de listas simples.

Normalización: Se aplicaron encabezados estandarizados (sin espacios y con mayúsculas iniciales) en los tres DataFrames.

###Resultados de Ventas Mensuales
Tras la limpieza, el script de cálculo arrojó una facturación anual sólida, con los siguientes hitos:

Mes de mayor desempeño: **May ($143,727.25)**

Mes de menor desempeño: **June ($108,480.17)**

###**Promedio de facturación mensual: Aprox. $122,000.00**

##**Conclusión Final del Proceso**
Los datos han sido validados y están libres de duplicados y nulos. La integridad de los tipos de datos garantiza que las fases posteriores de análisis (correlación entre marketing y ventas o segmentación de clientes por ingresos) se realicen sobre una base verídica.